From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen <haizhongdelan@outlook.com>
Date: Thu, 20 Jul 2023 19:59:12 +0800
Subject: [PATCH] Toggle sand duplication


diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 34b3541603b8cca16c7d62f3981d7ce3e8be0dbe..183619242765efd25b0cf3c1f217d80cf9a3f043 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -133,7 +133,7 @@ public class FallingBlockEntity extends Entity {
     @Override
     public void tick() {
         // Paper start - fix sand duping
-        if (this.isRemoved()) {
+        if (!this.level().atlantaConfig.sandDuplication && this.isRemoved()) { // Atlanta - Toggle sand duplication
             return;
         }
         // Paper end - fix sand duping
@@ -150,7 +150,7 @@ public class FallingBlockEntity extends Entity {
             this.move(MoverType.SELF, this.getDeltaMovement());
 
             // Paper start - fix sand duping
-            if (this.isRemoved()) {
+            if (!this.level().atlantaConfig.sandDuplication && this.isRemoved()) { // Atlanta - Toggle sand duplication
                 return;
             }
             // Paper end - fix sand duping
diff --git a/src/main/java/org/skydom/chosen/server/AtlantaWorldConfig.java b/src/main/java/org/skydom/chosen/server/AtlantaWorldConfig.java
index e50eeeffd3e6ba608166e10dddaf7dc4dc54eb6a..bd1a89daceeaa62e74c741b04c787ffc2cff3042 100644
--- a/src/main/java/org/skydom/chosen/server/AtlantaWorldConfig.java
+++ b/src/main/java/org/skydom/chosen/server/AtlantaWorldConfig.java
@@ -120,4 +120,10 @@ public class AtlantaWorldConfig {
         final Map<String, Object> value = getMap("world-settings." + worldName + "." + path, null);
         return value.isEmpty() ? fallback : value;
     }
+
+    public boolean sandDuplication = false;
+
+    private void FixSettings() {
+        sandDuplication = getBoolean("Fix.Enable-sand-duplication", sandDuplication);
+    }
 }
